### PR TITLE
Limit client internals knowledge

### PR DESCRIPTION
**Pull request by ouchadam**
Only the `Download` is given to clients via the `isAllowedToDownload` method which contains a limited amount of information (instead of the collection of `DownloadInfo`s)

### CODE REVIEW

**Comment by xrigau** [comment]
so a Download can be a batch?

File: library/src/main/java/com/novoda/downloadmanager/lib/DownloadInfo.java [hunk 1]
```diff
@@ -480,7 +480,11 @@ public void updateStatus(int status) {
     }
 
     private boolean isClientReadyToDownload(DownloadBatch downloadBatch) {
-        return downloadClientReadyChecker.isAllowedToDownload(downloadBatch);
+        return downloadClientReadyChecker.isAllowedToDownload(marshallToDownload(downloadBatch));
+    }
+
+    private Download marshallToDownload(DownloadBatch downloadBatch) {
+        return new Download(downloadBatch.getBatchId(), downloadBatch.getCurrentSize(), downloadBatch.getTotalSize());
```
Comment (ronocod) [inline library/src/main/java/com/novoda/downloadmanager/lib/DownloadInfo.java]: This is really confusing me:

Here you're passing the ID of a batch in as the ID of this `Download` object, so what does an instance of the `Download` class represent? Until now we've been using the term *'a Download'* to represent the download of a specific file.

Comment (xrigau) [inline library/src/main/java/com/novoda/downloadmanager/lib/DownloadInfo.java]: I think a Download can be either one item or a batch (as in 1 item is a batch of just 1 item)

Comment (ouchadam) [inline library/src/main/java/com/novoda/downloadmanager/lib/DownloadInfo.java]: I see a download as a batch

Comment (ronocod) [inline library/src/main/java/com/novoda/downloadmanager/lib/DownloadInfo.java]: Ok, in that case `DownloadInfo` should probably be renamed to `FileDownloadInfo` or something similar.

Comment (blundell) [inline library/src/main/java/com/novoda/downloadmanager/lib/DownloadInfo.java]: nice idea. 

This is the public external representation that clients can touch and mould to their hearts content. It free's our other models up to be touched and moulded by us without caring about *puny clients*.

File: library/src/main/java/com/novoda/downloadmanager/lib/Download.java [hunk 2]
```diff
@@ -0,0 +1,26 @@
+package com.novoda.downloadmanager.lib;
```
Comment (blundell) [inline library/src/main/java/com/novoda/downloadmanager/lib/Download.java]: let's have this outside of the `lib` package that'll help with clarification

File: library/src/main/java/com/novoda/downloadmanager/lib/Download.java [hunk 3]
```diff
@@ -0,0 +1,26 @@
...
+
+public class Download {
```
Comment (blundell) [inline library/src/main/java/com/novoda/downloadmanager/lib/Download.java]: maybe `DownloadPart` if it has a current size? MIXING IT UP

File: library/src/main/java/com/novoda/downloadmanager/lib/DownloadBatch.java [hunk 4]
```diff
@@ -2,7 +2,7 @@
 
 import java.util.List;
 
-public class DownloadBatch {
+class DownloadBatch {
```
Comment (blundell) [inline library/src/main/java/com/novoda/downloadmanager/lib/DownloadBatch.java]: yeah bai :100:

File: library/src/main/java/com/novoda/downloadmanager/lib/DownloadClientReadyChecker.java [hunk 5]
```diff
@@ -6,9 +6,9 @@
     /**
      * This method is executed on a background thread
      *
-     * @param downloadBatch
+     * @param download
```
Comment (blundell) [inline library/src/main/java/com/novoda/downloadmanager/lib/DownloadClientReadyChecker.java]: delete this line or fill in jdoc!

File: library/src/main/java/com/novoda/downloadmanager/lib/DownloadInfo.java [hunk 6]
```diff
@@ -480,7 +480,11 @@ public void updateStatus(int status) {
     }
 
     private boolean isClientReadyToDownload(DownloadBatch downloadBatch) {
-        return downloadClientReadyChecker.isAllowedToDownload(downloadBatch);
+        return downloadClientReadyChecker.isAllowedToDownload(marshallToDownload(downloadBatch));
```
Comment (blundell) [inline library/src/main/java/com/novoda/downloadmanager/lib/DownloadInfo.java]: we could have a marshaller class and call it `PublicFacingDownloadMarshaller` that'd clear things up (and improve testability)

File: library/src/main/java/com/novoda/downloadmanager/lib/DownloadClientReadyChecker.java (hunk 5 above)
Comment (stefanhoth) [inline library/src/main/java/com/novoda/downloadmanager/lib/DownloadClientReadyChecker.java]: image

**Comment by xrigau** [comment]
what's the deal with this PR, it's been open for 4 days now BE AGILE PLEASE

**Comment by xrigau** [comment]
sorry I forgot a smiley face :smile:

**Comment by ouchadam** [comment]
let's close it for now, there's a few changes to make

**Comment by hal9002** [comment]
Refer to this link for build results (access rights to CI server needed): 
https://ci.novoda.com/job/download-manager-prb/34/
Test PASSed.

**Comment by blundell** [comment]
This is a high priority change, the current public API we have is BAD

File: library/src/main/java/com/novoda/downloadmanager/lib/Download.java (hunk 2 above)
Comment (ouchadam) [inline library/src/main/java/com/novoda/downloadmanager/lib/Download.java]: :+1:

File: library/src/main/java/com/novoda/downloadmanager/lib/Download.java (hunk 3 above)
Comment (ouchadam) [inline library/src/main/java/com/novoda/downloadmanager/lib/Download.java]: mmmmmmmmmmmmmmmmmmmmmmmmm Download 4lyf

**Comment by hal9002** [comment]
Refer to this link for build results (access rights to CI server needed): 
https://ci.novoda.com/job/download-manager-prb/40/
Test PASSed.

**Comment by blundell** [comment]
LGTM :+1: any more :eyes:

File: library/src/main/java/com/novoda/downloadmanager/lib/DownloadThread.java [hunk 7]
```diff
@@ -729,10 +729,10 @@ private void readResponseHeaders(State state, HttpURLConnection conn) throws Sto
         }
 
         state.totalBytes = state.contentLength;
-        downloadInfo.setTotalBytes(state.contentLength);
+        fileDownloadInfo.setTotalBytes(state.contentLength);
 
         final boolean noSizeInfo = state.contentLength == -1 && (transferEncoding == null || !transferEncoding.equalsIgnoreCase("chunked"));
-        if (!downloadInfo.isNoIntegrity() && noSizeInfo) {
+        if (!fileDownloadInfo.isNoIntegrity() && noSizeInfo) {
```
Comment (amlcurran) [inline library/src/main/java/com/novoda/downloadmanager/lib/DownloadThread.java]: isNoIntegrity wat

**Comment by amlcurran** [comment]
LGTM

**Pull request merged**